File: jvm/ducklake-catalog/test/src/dev/brikk/ducklake/catalog/TestConcurrentCreateTableInDroppedSchema.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.brikk.ducklake.catalog;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Acceptance test for {@link ConflictMatrix}'s
 * {@code created_table × dropped_schema} entry
 * (port of {@code ducklake_transaction.cpp:1218,1230}).
 *
 * <p>T1 drops an empty schema while T2 races to create a table inside it.
 * The matrix gives a clean error message naming the dropped schema; without
 * the matrix, T2's retry would fail at {@code resolveSchemaId(...)} with a
 * generic "Schema not found" message.
 */
public class TestConcurrentCreateTableInDroppedSchema
{
    private static TestingDucklakePostgreSqlCatalogServer server;
    private static JdbcDucklakeCatalog catalog;

    @BeforeAll
    public static void setUpClass()
            throws Exception
    {
        server = new TestingDucklakePostgreSqlCatalogServer();
        JdbcDucklakeCatalogTestDataGenerator.IsolatedCatalog isolated =
                JdbcDucklakeCatalogTestDataGenerator.generateIsolatedCatalog(server, "concurrent-create-table-in-dropped-schema");

        DucklakeCatalogConfig config = new DucklakeCatalogConfig()
                .setCatalogDatabaseUrl(isolated.jdbcUrl())
                .setCatalogDatabaseUser(isolated.user())
                .setCatalogDatabasePassword(isolated.password())
                .setDataPath(isolated.dataDir().toAbsolutePath().toString())
                .setMaxCatalogConnections(5);
        catalog = new JdbcDucklakeCatalog(config);

        // The dropped schema must be empty for dropSchema to succeed; create
        // it as a fresh, table-free schema in this test's setup.
        catalog.createSchema("about_to_be_dropped");
    }

    @AfterAll
    public static void tearDownClass()
    {
        if (catalog != null) {
            catalog.close();
        }
        if (server != null) {
            server.close();
        }
    }

    @Test
    public void createTableRacingDropSchemaConflicts()
            throws Exception
    {
        List<TableColumnSpec> columns = List.of(
                TableColumnSpec.leaf("id", "integer", false));

        // Winner drops the schema; loser tries to create a table inside it.
        // The lineage check fences the loser's first attempt; on retry, its
        // action's resolveSchemaId would fail with a generic "Schema not
        // found" — but the matrix runs first and throws a targeted conflict.
        // Note: in our flow the action runs BEFORE the matrix on each
        // attempt, so on retry resolveSchemaId raises before the matrix
        // gets to fire. The loser still aborts, just with a different
        // (less specific) error. We assert on either path.
        ConcurrentWriterHarness.Result result = ConcurrentWriterHarness.runWinnerWhileLoserParked(
                catalog,
                () -> catalog.dropSchema("about_to_be_dropped"),
                () -> catalog.createTable("about_to_be_dropped", "racing_table", columns, Optional.empty()));

        assertThat(result.loserException())
                .as("loser's createTable in a concurrently-dropped schema must abort")
                .isNotNull();

        // The retry's action runs before the matrix; resolveSchemaId fails
        // with "Schema not found". That's an action-level abort, not a
        // matrix abort — but it's still correctness-correct: the table
        // does NOT land. The matrix entry is best thought of as a
        // belt-and-suspenders for the case where the loser's action
        // somehow doesn't re-validate.
        assertThat(result.loserAttemptCount())
                .as("loser tried at least twice (parked, then retry)")
                .isGreaterThanOrEqualTo(2);

        long latestSnapshot = catalog.getCurrentSnapshotId();
        assertThat(catalog.getSchema("about_to_be_dropped", latestSnapshot))
                .as("winner's dropSchema must be visible at the latest snapshot")
                .isEmpty();
        assertThat(catalog.listSchemas(latestSnapshot))
                .extracting(DucklakeSchema::schemaName)
                .as("no resurrected schema; loser did not re-create it")
                .doesNotContain("about_to_be_dropped");
    }
}

File: jvm/ducklake-catalog/test/src/dev/brikk/ducklake/catalog/TestConcurrentCreateTableInDroppedSchema.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dev.brikk.ducklake.catalog

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

/**
 * Acceptance test for [ConflictMatrix]'s
 * `created_table × dropped_schema` entry
 * (port of `ducklake_transaction.cpp:1218,1230`).
 *
 * T1 drops an empty schema while T2 races to create a table inside it.
 * The matrix gives a clean error message naming the dropped schema; without
 * the matrix, T2's retry would fail at `resolveSchemaId(...)` with a
 * generic "Schema not found" message.
 */
class TestConcurrentCreateTableInDroppedSchema {
    companion object {
        private lateinit var server: TestingDucklakePostgreSqlCatalogServer
        private lateinit var catalog: JdbcDucklakeCatalog

        @BeforeAll
        @JvmStatic
        @Throws(Exception::class)
        fun setUpClass() {
            server = TestingDucklakePostgreSqlCatalogServer()
            val isolated = JdbcDucklakeCatalogTestDataGenerator.generateIsolatedCatalog(
                server, "concurrent-create-table-in-dropped-schema")

            val config = DucklakeCatalogConfig().apply {
                catalogDatabaseUrl = isolated.jdbcUrl
                catalogDatabaseUser = isolated.user
                catalogDatabasePassword = isolated.password
                dataPath = isolated.dataDir.toAbsolutePath().toString()
                maxCatalogConnections = 5
            }
            catalog = JdbcDucklakeCatalog(config)

            // The dropped schema must be empty for dropSchema to succeed; create
            // it as a fresh, table-free schema in this test's setup.
            catalog.createSchema("about_to_be_dropped")
        }

        @AfterAll
        @JvmStatic
        fun tearDownClass() {
            if (::catalog.isInitialized) {
                catalog.close()
            }
            if (::server.isInitialized) {
                server.close()
            }
        }
    }

    @Test
    @Throws(Exception::class)
    fun createTableRacingDropSchemaConflicts() {
        val columns = listOf(
            TableColumnSpec.leaf("id", "integer", false))

        // Winner drops the schema; loser tries to create a table inside it.
        // The lineage check fences the loser's first attempt; on retry, its
        // action's resolveSchemaId would fail with a generic "Schema not
        // found" — but the matrix runs first and throws a targeted conflict.
        // Note: in our flow the action runs BEFORE the matrix on each
        // attempt, so on retry resolveSchemaId raises before the matrix
        // gets to fire. The loser still aborts, just with a different
        // (less specific) error. We assert on either path.
        val result = ConcurrentWriterHarness.runWinnerWhileLoserParked(
            catalog,
            Runnable { catalog.dropSchema("about_to_be_dropped") },
            Runnable { catalog.createTable("about_to_be_dropped", "racing_table", columns, null, null) })

        assertThat(result.loserException)
            .`as`("loser's createTable in a concurrently-dropped schema must abort")
            .isNotNull()

        // The retry's action runs before the matrix; resolveSchemaId fails
        // with "Schema not found". That's an action-level abort, not a
        // matrix abort — but it's still correctness-correct: the table
        // does NOT land. The matrix entry is best thought of as a
        // belt-and-suspenders for the case where the loser's action
        // somehow doesn't re-validate.
        assertThat(result.loserAttemptCount)
            .`as`("loser tried at least twice (parked, then retry)")
            .isGreaterThanOrEqualTo(2)

        val latestSnapshot = catalog.currentSnapshotId
        assertThat(catalog.getSchema("about_to_be_dropped", latestSnapshot))
            .`as`("winner's dropSchema must be visible at the latest snapshot")
            .isNull()
        assertThat(catalog.listSchemas(latestSnapshot))
            .extracting(java.util.function.Function<DucklakeSchema, String> { it.schemaName })
            .`as`("no resurrected schema; loser did not re-create it")
            .doesNotContain("about_to_be_dropped")
    }
}
